Webserver min memory

s0170071 opened the pull request:
Using 2.4.0 / lwip2 web interface becomes unresponsive now and then. Most likely reason is that free
memory went as low as 3k. Added a wait loop to wait for it to return ...
Fixes the issue as far as I can see.

EDIT: 
current PR keeps the free memory (no devices defined) always above 12k.  
The unresponsiveness is an issue of the 2.4.0 core and my Wifi repeater.

s0170071 committed:
Webserver min memory

memory went as low as 3k, so wait for it to return ...

Grovkillen commented:
This is pretty awesome if it is indeed fixing that issue. Good job! :+1:

s0170071 committed:
Adjusted to min 7k free

s0170071 commented:
Thanks. I'll improve it further adding a timeout. Otherwise it may stall if the core changes...

TD-er commented:
Hmm will test later this evening.
Lets hope it is indeed the solution.
Sounds simple enough.

And it will prevent watchdog resets ;)

Maybe we should also add some counter to not wait forever. What if the low memory is caused by the webserver allocating all those strings?
The watchdog won't step in to restart.

s0170071 commented:
maybe that while loop should go into the checkRam function with a parameter: 
checkRAM(F("sendWebPageChunkedData"), minFree);

s0170071 committed:
Timeout 100ms

TD-er commented on src/WebServer.ino:
```diff
@@ -300,7 +300,10 @@ void sendWebPageChunkedBegin(String& log)
 {
-  while (ESP.getFreeHeap() < 7000) backgroundtasks(); 
+  long timeout= millis();  
+  while ((ESP.getFreeHeap() < 7000) && ((millis()-timeout < 100)) ) {
```
Use something like this to wait for timeout:
```
uint32_t beginWait = millis();
while (!timeOutReached(beginWait + 1000)) {
```
It is better readable and does not cause issues when millis() wrap.

s0170071 committed:
tcp_cleanup for LWIP1 & 2

s0170071 committed:
Adapted timeout

s0170071 commented:
Hmmm. Works pretty well, however, if I let the ESP alone for a while, it seems to miss the first receiving packet. After that- fine again.

s0170071 commented:
And I have no Idea why Travis fails since https://github.com/letscontrolit/ESPEasy/pull/829/commits/1cbc49021a2e42d71687c9f8a4c81ada61e98179 . Compiles fine in Arduino IDE

TD-er commented:
It is being used here:
https://github.com/letscontrolit/ESPEasy/blob/58bb62cc1ae1873ed2be4b89d47e56e72342e47a/src/ESPEasy.ino#L1609-L1611
But as you can see, it is wrapped in the same ifdef checks.
So I am also not sure why it fails too, but I guess "ESP8266" is no longer used in the PlatformIO.ini.

ESP8285 is being used, but ESP8266 is not being set as define.

N.B. This may also affect some libraries like PubSubClient and Adafruit NeoPixel.

s0170071 commented:
But define esp8266 is used in various places in the code... 
It may also be that it failed because I made that last commit while the previous one was still compiling...

s0170071 committed:
trigger travis re-compile

trigger travis re-compile to see if the compile failure was due to a quick re-submission.

s0170071 commented:
@TD-er have you tested the PR ? 
Can you confirm that
- when it serves pages, you cannot crash it any more.
- it "hangs" when let alone for a minute or two ?

Can you ask your personal wire shark if 
- yours doesn't hang :-O
- the esp is silently dropping a packet or
- the packet is wrong or not sent due to some caching issue ? 

I'll fix that travis issue tonight. Maybe test the webserver example to see it it hangs too.

TD-er commented:
I will test it this evening.

s0170071 committed:
try fix travis

s0170071 committed:
Merge remote-tracking branch 's0170071/webserverminmem' into webserverminmem

s0170071 committed:
Does Travis like extern "C" ?

s0170071 commented:
Travis did not like the comment with the http:// link in it.

TD-er commented:
I'm still wondering who is going to set tcp_tw_pcbs.
It isn't initialized to null in the code.

s0170071 commented:
It's extern. Must be in the library.

TD-er commented:
I know, but my point is more like are we sure it always set to the correct value?
Thus initialized at 0 when no TCP thing to abort and set to some proper value that tcp_abort can do something with? (not a random uninitialized value)

TD-er commented:
Hmmm, running with your patch on 2.4.0 core.
It's not running well. 
I have 10k free and it keeps breaking http packets.
After saving the OLED Framed plugin, it was now showing something, but it was only a fraction of the web page. (before it was showing nothing, only some chunked error in Chrome)

So there's something not right here.
And I think I will change the view of the WiFi status on the OLED to reflect actual wifi status.
Will make a change + PR for that this evening, to ease diagnosing WiFi issues.

It also seems to be loading a lot longer on larger pages like the Info or the OLED framed page.

psy0rz commented:
hmm i dont quite understand this: If there is less memory available, it will process backgroundtasks, and this somehow magically free's memory?

s0170071 commented:
well, if you keep throwing strings at it (the send method) faster than it can send packets out, they get buffered. That's whats blocking the memory. If you wait for the packet to get sent out the memory gets freed eventually.

s0170071 commented:
@TD-er I will try something else. I wait now for ? 7k to be free. Try another, lower value. I'll try make that "must be free" memory a bit more dynamic.

TD-er commented:
Maybe try it a few times to get it lower. If that doesn't work, stop trying?

s0170071 committed:
Wait for memory be like before send chunked

goes not below 9680 bytes now

s0170071 commented:
just committed. It now looks what amount of memory is available before send and waits for that amount to show up afterwards. Just tried, works on my end and is a bit more "logical"

TD-er commented:
Shouldn't you release the memory first (`data = F("");`) and then perform the while loop?

s0170071 committed:
"Content-Type","text/html"

s0170071 commented:
nope. I register the amount of memory immediately before send. And wait for that to return. As an indication the packet is away if you want. Freeing anything else is an extra.

TD-er commented:
Tested with your last commit and it is working even worse.
Even the Tools page is not loading anymore, after only clicking around a few times.

s0170071 commented:
I don't understand this. Why is it working on my end ? I use 2.4.0 and Arduino IDE @ linux with Firefox. 
Whats your RSSI ? Can you try another router ? Any Firewalls ?

s0170071 commented:
Do you have the CRC in place ? Maybe your ESP Flash is worn ...

s0170071 commented:
Ahhh. And a very popular mistake: Flash with max 200kbps and set the CPU clock to 80MHz

TD-er commented:
Only "firewall" kind of thing here could be F-secure.
Does the CRC check also look after the settings part?

s0170071 commented:
Yes. It checks progmem and settings. The results are printed out to serial. As for the router, I've had issues with Netgear routers before. Yours isn't Netgear, is it?
I get the feeling your chunking issue is linked to this somehow.
Maybe someone else can try this PR ?

TD-er commented:
I only have Fritzbox stuff here.
And my Chromecast's should be patched by now ;) (they should be at 5 GHz and not affecting the ESP's)

I will try to lower the flash speed (serial is at 460800) and indeed, the ESP board here had some flash cycles to process. I've been using it now for 4 months I guess, continuous development.

s0170071 commented:
Fritzbox is fine, got one myself. Do you have a spare ESP ?
Dont't know if it matters but I do not have any devices set up. Can you clean your settings ?

TD-er commented:
Installed it on a blank fresh new NodeMCU board.
With Chrome I run into chunk errors in no-time and then it isn't even available for other browsers (firefox)
After a reboot I can connect to it with Firefox, but I see a lot of these:
```
WD   : Uptime 2 ConnectFailures 0 FreeMem 16056
IP blocked: 0.0.0.0 Allowed: 192.168.1.0 - 192.168.1.255
IP blocked: 0.0.0.0 Allowed: 192.168.1.0 - 192.168.1.255
IP blocked: 0.0.0.0 Allowed: 192.168.1.0 - 192.168.1.255
WD   : Uptime 3 ConnectFailures 0 FreeMem 17112
```

Will have to test with Wireshark to see what's going on here.

The ESP is reacting not a quick as I remember it was when trying the 2.4.0 libs for the first time.

s0170071 commented:
I've never had any block messages. Are they debug level 4? Do you have the CRC set up?
What happens in AP mode if you access it with your mobile?

TD-er commented:
I have set the access to the wifi, so it has a subnet from my DHCP server.
MD5 check says fail, but that's probably got to do with me not 'signing' the firmware with your python script.
No CRC errors on settings.

TD-er commented:
The new board (without anything else configured apart from WiFi) is now running for over an hour and it reacts quite fast.
Just as if some things had to settle somehow (there shouldn't be).
Saving to OLED framed (just created) is still a problem. Both via Firefox as well as Chrome.

s0170071 commented:
Nice. Could you pinpoint the source of your problems? Was it the upload speed?

Believe it or not, I had to swap the usb cable and the port once  to get it running.
Will try oled framed save in the morning.

s0170071 commented:
Ok, just tested. I can set up the framed display without problems. Memory goes as low as 6k but it works.

Lets talk about your power supply. Please do the following: 
- attach a short USP Cable. Try other cables and USB ports.
- do not use a bead board for supply.
- Solder a approx. 1uF ceramic capacitor right across the VCC / GND pins.
- remove (if any) loads from your NodeMcu
- make sure the reset pin is tied to 3v3. Don't leave it floating !

TD-er commented:
I always use decent USB cables, since the cheap ones are always causing issues.
I have it plugged in to an USB-3 port, which should be capable of 900 mA instead of 500.

Why ceramics cap and not an elco? Not sure if I have a ceramics capacitor at stock, bigger than 100 nF.
But will try this evening. The ESPs are powered via the on-board regulator.

The 2nd test setup, was a brand new, just opened NodeMCU with nothing else connected.

Where on the NodeMCU should I add a pull-up resistor to the 3v3? As close to the reset pin itself? Isn't there one already on the NodeMCU board? What do you suggest, 10kOhm?

s0170071 committed:
Optimized memory usage

memory never goes below 10k now. 
Added wait for memory return on header send, 
Added wait for memory return on chunk end send,
removed duplicate of data stashing with  getWebPageTemplateVar

s0170071 committed:
Merge remote-tracking branch 's0170071/webserverminmem' into webserverminmem

s0170071 commented:
Actually, with no devices defined and every main menu visited once, the memory is: 

![image](https://user-images.githubusercontent.com/33855380/36027424-81faef4c-0d9b-11e8-87f0-3d3e8c8f5d3a.png)

Theoretically you could run the web server with as little as 500-1000 bytes of ram. "just" a matter of programming. But Free Mem: 19944 (12528 - sendWebPageChunkedData) will do for now.

TD-er commented:
Just a comment on `pageTemplate =  F("");   //free mem -- s0170071- cannot confirm this`
String will allocate on the heap. How much is allocated, is stored in the capacity of the string.
By setting to an empty string, the capacity is probably not changed.
The only way to free the memory is to destruct the String.

I guess it is best to try to get some proper default value for the max. capacity of those strings and always reserve that amount (String::reserve() )
That way you always have the same block allocation on the heap, which may prevent heap fragmentation.

String is really awful on the heap.
When the capacity has to be increased, a new block will be allocated, data copied to the new block and then the old block is destructed.
Depending on the implementation, this increase may be just doubling the previous value.
So to store 1 byte more than the current capacity, means you need 3 times the current capacity as temporary heap allocation.

So reserve as soon as possible :)

s0170071 commented:
Let me know once you have tested it. Works fine for me. Without framed display.
As for your reset resistor: The NodeMcu has one already. So you're good. Try  plugging the ESP directly to the USB port (without hub)
How many FritzBoxes do you have ? I just found out this morning, that my second FritzBox (operated as Wifi Repeater) interfered with the ESP. If I let it alone for a minute, it(?) dropped the Wifi connection the moment I accessed it.

TD-er commented:
I have 2 Fritzbox repeater 1750E's (and one still in the box, brand new, read somewhere newer ones may cause issues)
Both repeaters are being setup as accesspoint, not repeater.
And a FB7581 modem.
The modem does not have wifi active at the moment, since the first firmware versions of that modem were unstable when WiFi and/or DECT was enabled.

I do have 2 Chromecasts, which I still have to check whether they have received the latest updates (there are reports of issues with them regarding WiFi)
But all the other ESP's run fine without any drop in WiFi connection.
Some even for > 30 days.

I still have to make the OLED display update the WiFi status (immediately) when connection is dropped.

s0170071 commented:
You may not notice that wifi drop. It appears as 7-15 s lag. Wifi is reestablished immediately and the tcp retransmit does the rest.

ShardanX commented:
@TD-er 
"Why ceramics cap and not an elco?"
Elco's usually are made from two aluminium foils on a paper wound up in a coil. This makes them worthless for high frequencies as the winding works more as a coil then a capacitor. Technically spoken they are too slow. As most integrated circuits produce spikes, extremely short pulses on the power supply line you will find a small ceramic capacitor near most of these IC's on a board.

From my experience i recommend capacitors from the X5R or X7R MLCC series even they are somewhat expensive. I'm using these on all my honmebrew circuits with best results. They are availlable up to 100µF usually at electronic dealers or Ali.  Some manufacturers offer even bigger ones, Murata for example offers up to 330µF but I've never seen them on public market.

s0170071 commented:
never mind, I am about to finish a web server version that is using no more than 2k of Ram. 
(Mind the minimum memory indicator :-)) After clicking all main menus.

![image](https://user-images.githubusercontent.com/33855380/36075935-b0b7a9b8-0f55-11e8-83e6-3e1a79e1b446.png)

Just need some more testing with the different core versions...

@TD-er, what about your chunking issue ? Did you try a plain vanilla wlan without repeaters ?

TD-er commented:
Not yet tested with different wlan (the fritzbox router)
Note that the repeaters only have that in their name. They do not act as repeaters, but are configured as accesspoints.
